### ISSUE TITLE
Two configs: include/exclude not overriding?

### ISSUE THREAD

**Issue by jimniels** [jekyll/jekyll#3746]
TLDR; When passing in two config files, the second file is not overriding excludes specified in the first file.

---

I have two config files for a single project:
- `_config.yml`
- `subfolder/_config.yml`

In the first `_config.yml` I have something like:

```
exclude:
    - subfolder
    - subfolder2
    - subfolder3
```

In the other config `subfolder/_config.yml` I have something like:

```
include: 
    - subfolder
```

When I pass in both configs, i.e. `jekyll serve --config _config.yml,subfolder/_config.yml`, the `subfolder` included in the second config doesn't override the `exclude` specified in the first config, and therefore doesn't get put into the output (`_site`)

**EDIT**: Overriding excluded folders doesn't appear to work whether or not you have two config files. i.e.:

```
exclude:
    - subfolder
    - subfolder2
    - subfolder3
include: 
    - subfolder
```

In this example configuration, `subfolder` would not be included, even though it is specified after the exclude

**Comment by envygeeks** [comment]
@benbalter would this get in the way of Github Pages?

**Comment by jimniels** [comment]
Note: It appears that having the same configuration attribute in two different files doesn't maintain the attributes set in the first. For example:

File 1: `_config.yml`
```
exclude: README.md
```

File 1: `subfolder/_config.yml`
```
exclude: subfolder
```

If I pass in both of these config files, Jekyll will *only* exclude `subfolder`. The fact that I specified `README.md` in the first file gets overwritten or lost.

**Comment by benbalter** [comment]
>  would this get in the way of Github Pages?

Nope. Thanks for checking in. We no longer use the command line multiple config option in production (although I personally use it all the time for development and test).

**Comment by jimniels** [comment]
Ran into this problem again. Have the need to exclude items across configs, but any excludes set in the first config get overwritten by whatever is being set in the second, rather than "in addition to".

**Comment by parkr** [comment]
@jimniels All the code for this is in `Utils#deep_merge_hashes`. All you need to do is type check [this line](https://github.com/jekyll/jekyll/blob/f3b99ebda75ae61201529c29d6e11db482b9a080/lib/jekyll/utils.rb#L29) and do a `|` (union) if both are arrays.

**Comment by jimniels** [comment]
@parkr to be honest, I know absolutely nothing about Ruby. Though following your advice and looking at the code, I took a shot and turned this line:

    target[key] = other_hash[key]

Into this:

    if target[key].is_a? Array and other_hash[key].is_a? Array
      target[key] = target[key] | other_hash[key]
    else
      target[key] = other_hash[key]
    end

And that seemed to work ... though idk if that breaks anything elsewhere?

**Comment by parkr** [comment]
You nailed it!

@benbalter, I am worried this could cause significant issues in the future because arrays will be combined rather than overridden. Can you envision a place where this could be an issue that couldn't be solved?

**Comment by benbalter** [comment]
> arrays will be combined rather than overridden

I'm not sure that that would be an expected behavior, and is probably best handled by having the desired (combined) array in the override file. 

I can't think of a specific example from the Jekyll config where merging arrays would be a problem (if not confusing), but imagine if I had `servers: [a,b,c]` for "production" and wanted `servers: [a,b]` locally... there'd be no way to override the `c` if I'm understanding correctly.

**Comment by parkr** [comment]
@benbalter In that case, you could have the config files build up from the smallest set to the largest. That being said, I could see `servers: [prod-api10, prod-api11, prod-api12]` combined with the default of `localhost` and that would not be what you wish. But for whom would this be the case? Jekyll isn't great for server configurations ;)

**Comment by envygeeks** [comment]
I think the expected behavior for most any user (including those who might not know any programming) is that there will be no union and that unions are more explicit than implicit.  If I put in a list of servers to override I would expect my servers to be the most important (in the context you have given) rather than combined.  It's easier to force them to re-add what they wish to keep to override than it is to deal with 90% of the cases where they just want their list (well I'm fudging the statistics obviously but you get it.)

**Comment by parkr** [comment]
Yeah, that makes sense to me. Perhaps, then, it should be explicitly documented?

**Comment by alimony** [comment]
I'm seeing this bug even for non-array values. The output when running `jekyll serve --config _config.yml,_config-dev.yml` is this:

    Configuration file: _config.yml
    Configuration file: _config-dev.yml
                Source: myproject
           Destination: myproject/_site
     Incremental build: disabled. Enable with --incremental
          Generating... 
                        done in 2.172 seconds.
     Auto-regeneration: enabled for 'myproject'
    Configuration file: myproject/_config.yml
        Server address: http://127.0.0.1:4000/
      Server running... press ctrl-c to stop.

What happens that second time that causes the additional config file to be ignored?

**Comment by parkr** [comment]
@alimony I believe we fixed that with https://github.com/jekyll/jekyll/pull/4953

**Comment by alimony** [comment]
@parkr: Yes, that's exactly it, thanks! When do you think it will be released?

**Comment by parkr** [comment]
@alimony It will be released with v3.2. No firm time yet, but :soon:
